docs/proposals/improved-schema-for-services.md

```diff
@@ -0,0 +1,409 @@
+
+# Improved schema for operator backed services
+
+This proposal is about improving the experience for a user of the operator backed services. Currently a user doesn't know exaustively which fields are available in a CR as we depend on optional metadata present in CRDs, to improve this we are considering using the metadata available from the cluster about the CRDs.
+Getting that metadata from the cluster is challenging because a normal cluster user ( plain vanilla ) doesn't have access to that metadata so we are gonna follow a 3 tiered approach
+
+- check if the user has access to the CRD, if so then use that.
+- get the `swagger.json` from the openapi endpoint provided by kubernetes. a sample can be found here https://gist.github.com/girishramnani/f29949d5cb8c6547102776437e05ac19
+- finally if we cannot get the openapi data as well then use the `ClusterServiceVersion` to get the parameters.
+
+This change would affect multiple service commands and the changes are described briefly below -
+- `odo catalog describe service` should include the metadata fields with description so the users can provide then when doing `odo service create`
+- `odo catalog list service` shouldn't change much other then listing CRDs per operator
+- `odo service create` should take flags and dynamically fill the CRD structs - metadata would be used for validation
+- `odo service create --from-file` would be used by adapter team or whoever wants to provide the whole CRD themselves as a file. this feature is already present and working.
+
+# Implementation plan
+
+Note - Below I would mention `cluster` manytimes and that means both openshift and kuberenetes. If there is something specific to one or the other then I would mention it.
+
+Below is the step wise plan for implementing this feature - 
+
+## Getting the metadata
+
+### Admin Access/Access to CRD API path
+First we would consider the scenerio where the user has admin access or privileges to access the CRD API to the cluster.
+Cluster provides an api to get metadata for any CRD needed. URL - `<cluster-url>/api/kubernetes/apis/apiextensions.k8s.io/v1/customresourcedefinitions`. 
+
+Sample output - 
+https://gist.github.com/girishramnani/cbb4400e463efe89c13f1386e0788793
+
+We care about the `openAPIV3Schema` as that would be used to build the CRD struct. 
+
+### Kuberenetes cluster Swagger has the schema
+
+If the user doesn't have CRD access then we fetch try to fetch the same `openAPIV3Schema` from the cluster's `swagger.json` from the endpoint `<cluster-url>/api/kubernetes/openapi/v2`. This is a very large document as it holds all the definitions present on the cluster.
+
+So this needs to be cached and refreshed whenever a new operator is installed.
```

Comment:
How we will detect that it needs to be refreshed? Storing the hash of all installed operators and comparing it if it didn't change?

We also need to keep in mind that users might work with multiple clusters, so the cache needs to be per cluster.

Reply:
https://github.com/openshift/odo/pull/4751#issuecomment-848601517

Reply:
had a similar approach in my mind shared in the original comment

Reply:
hashing would reduce the amount of information, and it the cache meta can be stored in `.odo` folder?

Reply:
sorry, did not notice that comment before. 

What about the multiple clusters problem? Directory for each cluster based on its api url? But than there might be issues with minikube/crc, where you can delete and create new cluster and url remains the same. It would be nice if we could get some kind of id/uuid of the cluster

Reply:
we can use the `cluster: api-engint-openshift-com:443` i.e. url with port like how kube context does it.

Reply:
so `{"api-engint-openshift-com:443": "<hash>",....}` would be how the operator-listing-cache.json ( we can decide on the name ) could look like?

Reply:
added